### ISSUE TITLE
Learning given an initial hypothesis

### ISSUE THREAD

**Issue by omarzd** [LearnLib/learnlib#57]
I would like to inquire about a certain feature, to start learning given an initial hypothesis. This can be implemented in the LearnLib adapter/client/test driver/SuL wrapper (whichever is the term) such that it will simulate the real SuL by replying to queries according to the initial hypothesis. However, I want it implemented in LearnLib itself and would like to know what that requires. 
The idea is that on the first iteration, instead of building a hypothesis using membership queries, LL would use the given hypothesis and start doing equivalence queries and continue normally from there. Of course the given hypothesis must match the given input alphabet plus other prerequisites. The advantage of this can be seen in two cases. The first case is simply resuming learning from a previously reached hypothesis instead of from the very beginning. This is helpful in case for some reason the learning process is halted. The second case is more complex. Suppose we can represent some knowledge about the SuL as a hypothesis. We want to introduce that hypothesis to LearnLib as a conjecture that it will treat the same way, i.e. start by testing it using equivalence queries.

In short, I would like to know how easy it is to implement a feature in LearnLib where it starts learning given a hypothesis.

Thank you

**Comment by omarzd** [comment]
Thanks for the thorough answer. Points:
1. I have to provide the discrimination tree, which means providing distinguishing suffixes.
2. I need to provide the differentiation between upper part (spanning tree edges) and lower part (distinguishing suffixes) of the observation table. 
3. I shall try out `ResumableLearner`.
4. I can guide the equivalence oracle to use certain traces using a `SampleSetEQOracle`.

Great. OK. Now I haven't thought about how I can transform a hypothesis from a DFA (or a Mealy machine) to an observation table, but I assumed it was doable only because the opposite transofrmation is. You're saying that such transformation is however not doable.

Point 2 seems to me the same as point 1. If I do provide distinguishing suffixes, then I will certainly be able to provide the distinction between upper and lower parts of the observation table. But I'm not familiar with the concept of spanning tree in this context. But a spanning tree is not unique, which means we can't extract the exact upper part of the observation table. I will investigate this further. Mainly the question is how would I translate knowledge about a system into an observation table.

**Comment by mtf90** [comment]
@omarzd any progress on this? Would you still regard this as an issue of LearnLib?

**Comment by omarzd** [comment]
No. I decided to go with your suggestion of guiding the learning process using my own oracle wrapper. 

Another idea I got out of this is I can make a higher-level component that acts as a **director** and---instead of guiding the learning process of a single learning experiment---it would set up multiple experiments with different alphabets (perhaps different levels of abstraction of the same alphabet) with the goal of learning several models that can somehow be combined together afterwards.
